mma: prototype

sumeerbhola opened the pull request:
Informs #103320

Epic: CRDB-25222

Release note: None

wenyihu6 committed:
wh: rename regions to a/b/c in tests

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142138)
<!-- Reviewable:end -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142138)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/142138#-OKCTJ6E-74C8J3oU43Q:-OKCTJ6E-74C8J3oU43R:b-qp9boi) (no related file):*
The first commit is from https://github.com/cockroachdb/cockroach/pull/142002

@kvoli This depends on support in clusterState that will get added by your PR. The code here is tentative with many TODOs, some of which are trivial, some for post-protoyping, and some dependent on clarifying the semantics we want for diversity decrease. I am leaning towards this being the start of the prototyping branch since we are now starting to write non-data-structure code and much more is subject to change on how we do the integration.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142138)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/142138#-OKCTJ6E-74C8J3oU43Q:-OKCsDrW9PepHLz5Dxj8:bwwj7i4) (no related file):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The first commit is from https://github.com/cockroachdb/cockroach/pull/142002

@kvoli This depends on support in clusterState that will get added by your PR. The code here is tentative with many TODOs, some of which are trivial, some for post-protoyping, and some dependent on clarifying the semantics we want for diversity decrease. I am leaning towards this being the start of the prototyping branch since we are now starting to write non-data-structure code and much more is subject to change on how we do the integration.
</blockquote></details>

@kvoli added a commit for rebalancing leases. I'll stop here until we can discuss next steps.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142138)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/142138#-OKCTJ6E-74C8J3oU43Q:-OKCz7ul2ZX4LsDb8IJP:bdbfgz1) (no related file):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

@kvoli added a commit for rebalancing leases. I'll stop here until we can discuss next steps.
</blockquote></details>

Ack, lets discuss on Monday.


<!-- Sent from Reviewable.io -->

kvoli commented:
Rebased to resolve conflicts.

kvoli commented:
Fixed up the last commit to resolve shadowing a var.

kvoli commented on pkg/kv/kvserver/allocator/mma/allocator_state.go:
```diff
@@ -61,17 +94,267 @@ func (a *allocatorState) computeChanges() []*pendingReplicaChange {
+				if !a.ensureAnalyzedConstraints(rstate) {
+					continue
+				}
+				cands, _ := rstate.constraints.candidatesToMoveLease()
```
TODO: add in logic to strictly pick first preference which has >0 replicas. (move to todo in the code)

kvoli replied:
Added todo in `candidatesToMoveLease`

kvoli commented on pkg/kv/kvserver/allocator/mma/load.go:
```diff
@@ -309,10 +313,35 @@ func (mm *meansMemo) getMeans(expr constraintsDisj) *meansForStoreSet {
+	return summary
+}
+
+func computeMeansForStoreSet(
```
Add code TODO for excluding stores based on node failure detection from the mean.

kvoli commented on pkg/kv/kvserver/stores.go:
```diff
@@ -335,3 +335,17 @@ func (ls *Stores) GetStoreMetricRegistry(storeID roachpb.StoreID) *metric.Regist
+		nc.StoresCPURate += int64(c.CPUPerSecond)
+		return nil
+	})
+	// TODO: fill in the rest of the fields.
```
@kvoli take this.

kvoli committed:
asim: remove redundant range usage info on lease transfer

`RangeUsageInfo` was never used, remove it from `op.TransferLeaseOp`.

Informs #103320
Epic: CRDB-25222
Release note: None

kvoli committed:
mma: reset range analyzed constraints on store leaseholder msg

Also, add debug information and additional panics on inconsistent
states.

Informs #103320
Epic: CRDB-25222
Release note: None

kvoli committed:
mma,asim: sim integration

This commit integrates the mma.Allocator into `asim`. No changes are
currently returned when enabling the multi-metric store rebalancer via
`LBRebalancingMode=3` (see `testdata/non_rand/mma`).

In service integration:

- `cpu` is now recorded as a fixed cost in `state/load.go` on load
  events
- `Allocator.ProcessStoreLoadMsg` is called when receiving a gossip update,
  similar to the storepool
- Gossiped store descriptors are populated with `NodeCapacity` in
  `asim/gossip`
- `Allocator.SetStore` is called on every node's allocator when a new
  store is added to the simulator and also when a store receives an
  update via gossip.

Informs #103320
Epic: CRDB-25222
Release note: None

sumeerbhola committed:
mma: remove panic from IsTransferLease; initialize RaftCPU for asim

sumeerbhola committed:
mma: fix replicas consistency bug in clusterState.processStoreLeaseholderMsgInternal

sumeerbhola committed:
mma: cleanup some code and commentary related to roachpb.ReplicaType

kvoli committed:
mma: interpret localNodeID via localStoreID

`allocatorState` depends on knowing the local node ID to determine which
stores are local, in order shed leases. See comment:

``go
// NB: We interpret the local NodeID based on the given localStoreID because
// the allocator is initialized when starting a server, so the NodeID is not
// yet known.
``

Remove the `nodeID` from allocator state for now, instead interpret it from
the given `localStoreID` when calling `rebalanceStores`.

Informs #103320
Epic: CRDB-25222
Release note: None

sumeerbhola committed:
mma: change signature of computeMeansForStoreSet

kvoli committed:
mma: add log statements when shedding leases in rebalanceStores

Informs #103320
Epic: CRDB-25222
Release note: None

kvoli committed:
asim: add setting to disable lease and replicate queues

Informs #103320
Epic: CRDB-25222
Release note: None

kvoli committed:
asim: move mma -> mma_high_cpu

We will use this for testing cpu rebalancing using mma. Note that this
doesn't cover high cpu on follower stores necessarily.

Informs #103320
Epic: CRDB-25222
Release note: None

kvoli committed:
asim: add mma full disk simulation test

Run with:

```
dev test pkg/kv/kvserver/asim/tests -v -f TestDataDriven/mma_full_disk
```

Informs #103320
Epic: CRDB-25222
Release note: None

kvoli committed:
kvserver: pass rand when initializing mma.allocator

Informs #103320
Epic: CRDB-25222
Release note: None

kvoli committed:
mma: update rebanceStores log format to be more consistent

Epic: none
Release note: None

kvoli committed:
asim: fix range/store capacity write bytes per second and size

WriteBytesPerSecond wasn't being used for non-leaseholder replicas when
calculating their store's `Capacity`, fix this. Also, generally allow
calling `State.RangeUsageInfo` for non-leaseholder replicas, which
will be identical to the leaseholder replica, minus the
`QueriesPerSecond` and `RequestCPUNanosPerSecond`.

The delta in some test files is related to these changes.

Epic: none
Release note: None

sumeerbhola committed:
tests: tweaks to example_rebalancing

kvoli committed:
asim: allow specifying cluster capacity in TestDataDriven

Previously, the capacity needed to be set via `set_capacity capacity=x`.
Add a `store_byte_capacity=x` field to `gen_cluster` in `TestDataDriven`
which allows setting the store byte capacity when generating a cluster.

Use a default `256 GiB` `store_byte_capacity` when none is specified.
Note, this doesn't set a default capacity for other tests, like the
rand tests.

Epic: none
Release note: None

kvoli committed:
mma,asim: add debugging log statements for rebalance changes

These should be removed eventually. Run with:

```
dev test pkg/kv/kvserver/asim/tests -v -f TestDataDriven/mma_high_cpu --show-logs  --ignore-cache
```

Panic is on multiple voter additions, when there should only be
rebalancing and lease transfers:

```
store 1: computed 28 changes [r4=[change_replicas=[{ADD_VOTER n1,s1} {ADD_VOTER n7,s7}]]...
```

Epic: none
Release note: None

sumeerbhola committed:
mma: add some logging to understand test behavior

sumeerbhola committed:
mma: some more logging

sumeerbhola committed:
mmaintegration: some printfs related to inconsistency

kvoli committed:
asim: fix MakeStoreLeaseholderMsgFromState for only leaseholders

This was generating a range message if the replica was **not** a
leaseholder, rather than the inverse, when the store replicas is a
leaseholder. Fix this and convert the printf logging to panics.

Epic: none
Release note: None

kvoli committed:
mma: fix bug where lease transfers were considered rebalances

This was due to an error in `IsChangeReplicas` which incorrectly
considered lease add/remove as a replication change. Fix this and add
more logging.

Now hitting a bug in the simulator code, which cannot find the operation
associated with a pending ticket.

Epic: none
Release note: None

kvoli committed:
asim: don't remove pending operation tickets from controller

Tickets were being removed upon calling `Controller.Check()`, which led
to missing tickets. Unsure why we did it this way originally, perhaps to
save memory. In any case, its simpler to just leave all the tickets
around.

The mma high cpu datadriven test now runs e2e, applying changes.

Epic: none
Release note: None

kvoli committed:
asim: remove stale comment on mma_high_cpu

Epic: none
Release note: None

sumeerbhola committed:
mma: fix non-determinism in sheddingStores sort order

kvoli committed:
asim: add last store plot value and write_bytes_per_second

Epic: none
Release note: None

sumeerbhola committed:
mma: tweak loadSummaryForDimension so above mean becomes loadNoChange

sumeerbhola committed:
mma: fix sort ordering bug in sortTargetCandidateSetAndPick

also remove the sheddingThreshold logic in computeCandidatesForRange, since
I could not figure out what it was doing.

sumeerbhola committed:
mma: limit the number of changes produced in a rebalance pass

kvoli committed:
asim: add mma_high_cpu_skewed_placement datadriven simulation test

This test is similar to `mma_high_cpu` with two exceptions:
- uniform workload access distribution instead of zipfian
- initial replica placement is skewed (all leases on n1s1, replicas on
  n1s1 n2s2 n3s3)

Currently no rebalancing takes place, which should be investigated.

Epic: none
Release note: None

kvoli committed:
asim: allow disabling split queue

Epic: none
Release note: None

sumeerbhola committed:
mma: adjust logging and increase mma_high_cpu duration

sumeerbhola committed:
mma: canAddLoad prevents going above the mean

kvoli committed:
asim: allow specifying a node cpu capacity in dd tests

via `node_cpu_rate_capacity`, default is 0.

Epic: none
Release note: None

kvoli committed:
asim: fix bug where change replicas used rangeid as key

Epic: none
Release note: None

sumeerbhola committed:
mma: compute maxFractionPending

kvoli committed:
asim,mma: thread ctx and fix change replicas bug #2

The bug caused change replicas operations never to be marked as complete
in the `Controller`, causing the mma store rebalancers to stall after
their first change replicas.

Epic: none
Release note: None

sumeerbhola committed:
mma: fix logging typo

sumeerbhola committed:
mma: fix MakeStoreLoadMsg to compute a reasonable disk capacity for a near empty store

sumeerbhola committed:
mma: fix MakeStoreLoadMsg when cpu usage is close to zero

kvoli committed:
mma: fix store lease shedding loggiing

Epic: none
Release note: None

kvoli committed:
asim: extend mma_high_cpu_skewed_placement duration 5->25m

To see when rebalancing ends.

Epic: none
Release note: None

kvoli committed:
mma: add debug logging in can add load

Epic: none
Release note: None

sumeerbhola committed:
mma: make top-k deterministic for testing

kvoli committed:
mma: assert shedding range leaseholder is local store

Epic: none
Release note: None

kvoli committed:
mma: cleanup makePendingChanges and add lease transfer validation

Epic: none
Release note: None

kvoli committed:
asim: disable split queue in mma_high_cpu test

Epic: none
Release note: None

kvoli committed:
asim: actually disable the split queue via settings

Epic: none
Release note: None

kvoli committed:
asim: make load configurable over keyspace

Allow configuring the workload and range generation so that workloads
can be tied directly to range generation, via specifying the same
`min_key` and `max_key`.

The hardcoded request cpu per request and raft cpu per write are
replaced by the workload generator, which allows specifying
`request_cpu_per_access` and `raft_cpu_per_write`.

Epic: none
Release note: None

sumeerbhola committed:
mma: extend mma_high_cpu interval to 60min

We should be able to undo this extension once we are executing more
than 2 range movements per 60s.

kvoli committed:
kvserver: provide settings when setting up mma store rebalancer

This was causing a nil ptr panic.

Epic: none
Release note: None

kvoli committed:
asim: call allocator.ComputeChanges until no changes are returned

Update the `MMAStoreRebalancer` to call `ComputeChanges` until no changes
are returned. Previously, `ComputeChanges` was called exactly once
per `LBRebalancingInterval`.

Epic: none
Release note: None

sumeerbhola committed:
mma: two changes to the heuristic

- replicas below some threshold are ignored in the top-k. This prevents
  movement of replicas that are high in a dimension that is not overloaded
  from being shed.
- a store can accept replicas if it is less than 5% above the mean.

kvoli committed:
mma: favor local store when ordering shedding stores

Epic: none
Release note: None

kvoli committed:
asim: add lease_weights and replica_weights gen_ranges

When the `placement_type` is `weighted`, this allows specifying
`lease_weights=(...)` and `replica_weights=(...)` to directly control
the initial replica and lease distribution outputted from `gen_ranges`.

A datadriven test is added, using the weighted placement type in order
to test remote store replica shedding for CPU load.

Epic: none
Release note: None

sumeerbhola committed:
mma: tiny cleanup and shorten mma_high_cpu duration

sumeerbhola committed:
mma: remove early filtering based on load and maxFractionPending

sumeerbhola committed:
mma: add candidateInfo.leasePreferenceIndex

sumeerbhola committed:
mma: reduce maxRangeMoveCount to 1

sumeerbhola committed:
mma: introduce loadMultiplierForAddition=1.1

Fixes an existing todo. Also dampens the thrashing in
mma_high_cpu_skewed_placement.

sumeerbhola committed:
mma: tweak the logic for topk.dim

sumeerbhola committed:
mma: use leasePreferenceIndex when transferring lease and rebalancing range

sumeerbhola committed:
mma: delay filtering for lease shedding until sortTargetCandidateSetAndPick

sumeerbhola committed:
mma: relax canAddLoad to canShedAndAddLoad

Now it will return true if the post change load summaries of the src is
at least as overloaded as the target.

This is not the complete change to consider higher load stores as targets,
since we currently do not consider those higher load targets in
sortTargetCandidateSetAndPick. There will be another commit with that
change.

sumeerbhola committed:
mma: fix a bug in lease transfers related to source loadSummary

We were not evaluating whether the source was overloaded wrt the set of
leaseholder candidates. Fixing this caused problems with the existing
mma_high_cpu because the simulator sets a high cpu capacity for s1,
which causes the utilization to dip despite the high load, and which
results in s1 looking like loadLow. We now use the utilization only
to err on the side of deciding that a store is overloaded, by making
a change to loadSummaryForDimension. We should still consider this
method a temporary heuristic that is subject to many future changes.

sumeerbhola committed:
mma: remove a todo comment

kvoli committed:
asim: add node cpu rate capacity to mma_high_cpu

Epic: none
Release note: None

sumeerbhola committed:
mma: add partial ability to rebalance higher to lower overload stores

This is mostly disabled via tempOverrideToIgnoreLoadNoChangeAndHigher
(see comment on why).

sumeerbhola committed:
mma: introduce storeLoadSummary.dimSummary which fixes a bug and todo

wenyihu6 committed:
wh: fix dev gen

sumeerbhola committed:
mma: remove todos in messages.go

sumeerbhola committed:
mma: fix comments and todos in load.go

sumeerbhola committed:
mma: fix some commentary and todos in constraint.go and allocator_state.go

sumeerbhola committed:
mma: remove a cluster_state.go todo

sumeerbhola committed:
mma: remove an allocator_state.go todo

sumeerbhola committed:
mma: set overrideToIgnoreLoadNoChangeAndHigher to false for lease transfers

kvoli committed:
mma: add approach to avoid rebalancing replicas for remote stores early

See commentary.

Epic: none
Release note: None

kvoli committed:
asim: enable mma store rebalancer in mma_high_cpu_unable_to_shed_leases

To match the comment above the settings command.

Epic: none
Release note: None

wenyihu6 committed:
asim: mma pair programming

kvoli committed:
kvserver: add RegisterExternalChanges method to mma.Allocator

Add `RegisterExternalChanges` to the `mma.Allocator`, this method is
intended to be used for registering changes which are external to the
`mma.Allocator`. Also export `ReplicaChange` and `ChangeID` to support
registering these.

Epic: none
Release note: None

wenyihu6 committed:
wh: asim - add ability to specify node locality when generating a cluster setup

kvoli committed:
kvserver: add allocator integration skeleton

This is likely logically a bit flawed, but the core intent remains.

We need to register non-mma operations, get some changeIDs and then
enact them.

In the future, there should be a simulation and non-simulation execution
component, which uses the `AllocatorSync` to apply state updates. The
reason for the separation, is the different in execution model. The
production code blocks on its changes, while the simulation code has a
non-blocking model.

Epic: none
Release note: None

sumeerbhola committed:
mma: fix comment in mma_high_cpu_unable_to_shed_leases

s1 is the one with no leases, not s5.

sumeerbhola committed:
mma: tiny syntactic tweak to not reuse topKRanges and n

wenyihu6 committed:
Revert "wh: rename regions to a/b/c in tests"

This reverts commit 7e90f3e4e6e6c7d3a7641efc5b1931806b126731.

wenyihu6 committed:
asim: split the span config initialization

wenyihu6 committed:
asim: add directive to place things correctly for constraints

still need more tests and check

wenyihu6 committed:
asim: add more tests for restrictive placement

wenyihu6 commented:
Rebased to pick up https://github.com/cockroachdb/cockroach/commit/607b5e57b9aeaf08bdfe12c73084886287615f98.

wenyihu6 committed:
asim: change lease weights to lease

wenyihu6 committed:
asim: user needs to give a more restrictive input

kvoli committed:
kvserver: rm policy_refresher_test

Looked to be unintentionally included.

Epic: none
Release note: None

wenyihu6 committed:
asim: new input structure

wenyihu6 committed:
asim: new input structure

wenyihu6 committed:
asim: add SetSimulationSettingsEvent

wenyihu6 committed:
asim: fix nodes_per_region parameter to take in []int instead

wenyihu6 committed:
mma: test data

wenyihu6 committed:
mma: add load to mma_constraint_satisfaction1

wenyihu6 committed:
asim: change rebalance_mode to 3

wenyihu6 committed:
asim: add mma_constraint_satisfaction_old_alloc

wenyihu6 committed:
asim: add show_initial_value=true to plot

wenyihu6 committed:
asim: sort by storeID instead of value for orderedStoreIDs

sumeerbhola committed:
non_rand: add mma_constraint_satisfaction1_full_disk

It doesn't work as expected.

sumeerbhola committed:
mma,non_rand: reduce duration of mma_constraint_satisfaction1

sumeerbhola committed:
mma: don't move replica from local store if managed to produce at least 1 lease transfer

Instead we hope that by the time the rebalancer finishes the lease transfer,
it may be able to able to find another lease transfer (which is cheaper than
moving a replica).

sumeerbhola committed:
mma: only add CPURate load when transferring the lease

sumeerbhola committed:
mma: only consider target's CPURate summary when doing a lease transfer

wenyihu6 committed:
kvserver: add the defer I think it was missing

wenyihu6 committed:
asim: add defer I think the test was missing

wenyihu6 committed:
asim: fix mma_constraint_satisfaction1_full_disk

wenyihu6 committed:
mma: subsumesChange should be a pointer receiver method since it changes the replica type

kvoli committed:
mma,asim: reduce mean rebalance threshold 20%->10%

Epic: CRDB-25222
Release note: None

kvoli committed:
asim: move storepool to be per-node instead of per-store

There is one storepool shared among every store on a node, match this in
the simulator.

Epic: none
Release note: None

kvoli committed:
kvserver: notify storepool <-> mma of changes

Introduce an `allocator_integration` component, which is called into by
both the mma store rebalancer and the existing lease/replicate queues
(in the simulator) to notify the other component about pending changes.

TODO: Populate `mma.clusterState` with range information when the range
doesn't currently exist but is having an external change registered.

Epic: none
Release note: None

kvoli committed:
asim: add mma_high_write_uniform_cpu_all_enabled

Which is identical to `mma_high_write_uniform_cpu` but with the
replicate/lease/split queues enabled.

Epic: none
Release note: None

kvoli committed:
asim: use 256mb range size in mma_high_cpu_skewed_placement

Epic: none
Release note: None

sumeerbhola committed:
mma: be less aggressive about rebalancing based on ByteSize

We've seen instances where rebalancing due to WriteBandwidth gets
interference (resulting in thrashing) due to ByteSize becoming slightly
overloaded.

kvoli committed:
asim: fix snapshot rebalance rate to be per-second

The snapshot rate calculation incorrectly divided the range size by the
snapshot rate without considering the snapshot rate was per-second.

Fix this. This causes a large delta in some generated tests.

Epic: none
Release note: None

kvoli committed:
asim: increase snapshot rate to 32 mib/s

This matches the default in the code, previously it was 16 MiB/s.

Epic: none
Release note: None

sumeerbhola committed:
non_rand: tweak some test durations

sumeerbhola committed:
non_rand: add test that creates problems for alloc1 and mma

Next step is to understand why.

sumeerbhola committed:
mma: change the logic related to summarizing ByteSize

sumeerbhola committed:
non_rand: compare sma and mma in the absence of queues

sumeerbhola committed:
non_rand: add example_skewed_cpu_even_ranges_sma_and_queues

Shows thrashing because replicate and lease queues are unaware of cpu

sumeerbhola committed:
mma: remove stale comment

sumeerbhola committed:
non_rand,mma: add mma_skewed_cpu_skewed_write

And tweak the loadSummaryForDimension logic. This test strains the
ability of MMA to equalize both CPU and write-bandwidth. There are
todos to improve the behavior of this test.

kvoli committed:
asim: allow specifying rebalance_objective and cleanup settings

Remove some unused simulation settings and allow specifying
`rebalance_objective`. The following tests are updated to have a higher
CPU workload and the non-mma tests to use `rebalance_objective=1` (cpu).

```
example_skewed_cpu_even_ranges_mma
example_skewed_cpu_even_ranges_sma
example_skewed_cpu_even_ranges_sma_and_queues
```

Epic: none
Release note: None

sumeerbhola committed:
mma: change the logic in sortTargetCandidateSetAndPick

Specifically, rework the logic around
discardedCandsAtOrBeforeLowestLoadHadPendingChanges. Ordinarily,
we should work only with the set that has the lowest store loadSummary.
We want to make an exception when everything in that set has a high
node loadSummary. But the logic to only do this when there are no
stores in the earlier sets that have pending changes was flawed --
it is now fixed to be stricter. This should delay making suboptimal
choices.

sumeerbhola committed:
mma: tweak logging

sumeerbhola committed:
mma: modified approach to rebalancing target selection

This scheme allows us to rebalance when all stores are overloaded along
some dimension.

This is not yet complete, in that we need to change the value of
tempOverrideToIgnoreLoadNoChangeAndHigher.

sumeerbhola committed:
mma: reduce thrashing in clusterState.canShedAndAddLoad

This is done by considering the target and source also in the overload
dimension.

sumeerbhola committed:
logging for SMA

sumeerbhola committed:
mma: pass context to processStoreLoadMsg

kvoli committed:
kvserver: add todos/comments on allocator1 integration

Epic: none
Release note: None

sumeerbhola committed:
mma: don't shed from stores that have a pending increase

sumeerbhola committed:
mma: delay shedding replicas from remote store to allow for lease shedding

kvoli committed:
kvserver: record new per second cpu in gossip

Also, hook up the simulator to also call `RecordNewPerSecondStats`.

Epic: none
Release note: None

sumeerbhola committed:
mma: allow rebalancing between overloaded stores

This finally allows mma_skewed_cpu_skewed_write to successfully
rebalance both cpu and write bandwidth.

kvoli committed:
asim: correctly set queue next time

The asim lease/replicate queue wouldn't correctly wait for change
application, as they were not updating the next tick time correctly.

Epic: none
Release note: None

sumeerbhola committed:
asim: add example_skewed_cpu_even_ranges_mma_and_queues

kvoli committed:
kvserver: record follower write bytes in replica load

Previously, only bytes written at the leaseholder were accounted for on
a replica. Now, follower replicas will record the bytes written to their
replica load.

Epic: none
Release note: None

sumeerbhola committed:
mma: borrow the fix to NonMMAPreChangeReplicas regarding lack of info about leaseholder

The fix was in a commit by Austen
https://github.com/cockroachdb/cockroach/commit/8a5bea0303fad19d324fd9b4fc1735ac296e37a4
that was not added to the prototype since it included other changes that
broke aspects of the prototype.

sumeerbhola committed:
mma: don't transfer lease to an existing voter that is lagging

sumeerbhola committed:
mma: remove lastLeaseShedAt and related changes

We have changed to a purely time delay based approach for expanding the
set of choices for rebalancing, as expressed using
remoteStoreLeaseSheddingGraceDuration and related *GraceDurations.

sumeerbhola committed:
mma: simplify loadSummaryForDimension etc.

sumeerbhola committed:
mma: remove a todo in constraint.go and clarify why

sumeerbhola committed:
mma: remove storeChangeRateLimiter and helpers

We had hypothesized needing this to coordinate across rebalancers runnning
in different stores, but we never integrated it and it hasn't been necessary
for the asim runs.

sumeerbhola committed:
mma: fix some commentary in cluster_state.go

sumeerbhola committed:
mma: cleanup the pending changes logic

Except for the load pending changes, the other changes were supposed
to be removed when enacted. But we were delaying the removal until removal
of the load pending changes. This was not consistent with the commentary,
so this is fixed now. There is regression in some tests, notably the test
where MMA runs with all the queues enabled, possibly because MMA is acting
faster.

sumeerbhola committed:
mma: small comment tweaks in allocator_state.go

sumeerbhola committed:
mma: clarify the origin of the min{Lease,Replica}LoadFraction

sumeerbhola committed:
asim: add mma_high_cpu_25nodes

sumeerbhola committed:
asim: increase the write bandwidth for some tests

This is to prepare for a change that will discount small differences
in write bandwidth.

sumeerbhola committed:
mma: change loadSummaryForDimension to ignore small differences

- WriteBandwidth is scaled down to multiples of 1MiB/s when comparing so
  that small differences don't have consequences.
- CPURate that is < 5% utilization can never be considered overloaded, even
  if far away from the mean.

wenyihu6 committed:
kvserver/mma: do not create empty range state for RegisterExternalChanges

Previously, MMA would create an empty range state if RegisterExternalChanges
received stats for an unknown range. This patch changes that behavior: MMA now
ignores replica changes for unknown range IDs, and PostApply updates the store
pool but skips updating MMA for ranges that don't exist at the time of
application.

sumeerbhola committed:
mma: change the incremental behavior of StoreLeaseholderMsg

A RangeMsg is now included for all ranges for which the store is a
leaseholder. A Populated bit is set in the RangeMsg if all the details
are populated (if something changed), else only the RangeID is provided.
This allows MMA to find ranges for which the store was the leaseholder
and is no longer the leaseholder and remove them from its state.

wenyihu6 committed:
kvserver: only the first change id should be tracked in as.trackedChanges

sumeerbhola committed:
mma: fix stale construction of RangeUsageInfo

sumeerbhola committed:
mma: tiny comment and naming cleanup

sumeerbhola committed:
mma: fix bug (found by Wenyi) where we were re-applying the load change

This improved mma_high_write_uniform_cpu_all_enabled

sumeerbhola committed:
kvserver,mma: don't ask MMA to rebalance if it does not know of stores

This is a stop-gap given that mma.Allocator.SetStore is only being
called when a gossip update is received for a store, so the stores
knowledge of MMA can be stale.

Also did some other minor cleanup:
- Use the existing kv.allocator.load_based_rebalancing_interval to
  control the interval of rebalancing.
- Loop in the mma rebalancer until there are no more changes, as is
  done by the existing rebalancer.

sumeerbhola committed:
mma: add commentary related to rangeState

Including justifying why it is beneficial to guarantee at most once
rangeState per RangeID.

sumeerbhola committed:
mma: make AllocatorSync threadsafe

wenyihu6 committed:
kvserver: integrate allocator sync with non-mma components

TODO(wenyihu6): many parts feel inconsistent. Try looking at the responsibilities of each struct and refactor to clean up.

wenyihu6 committed:
kvserver: integrate AllocatorSync into mma

wenyihu6 committed:
kvserver: remove ApplyImpact

ApplyImpact is no longer needed since we update local store pool directly in PostApply (changeReplicas, TransferLease).

wenyihu6 committed:
mma: skip mma if there already is a pending range change

Skips RegisterExternalChanges if there already is a pending
range change. This made mma_high_write_uniform_cpu_all_enabled thrashing.
Need a closer look.

sumeerbhola committed:
mma: make Allocator impl thread safe

sumeerbhola committed:
mma: fix bug in AllocatorSync where StorePool was not being updated

Specifically, any replica moves made by MMA were not being recorded
correctly. Also updated PendingRangeChange.ReplicationChanges to handle
more cases.

wenyihu6 committed:
mma: addressed comments from sumeer on the previous commits

wenyihu6 committed:
mma: add more debug lines for replica change

We can revert this if it is not useful.

wenyihu6 committed:
mma: addressed sumeer's comment and cleaned up some obs lines

Previously, I misunderstood what sumeer's nit wanted in 77979c1672073b15c8c5a9ed8ff86b1d7751c699. This commit clears it up.

wenyihu6 committed:
mma: add some basic metrics for mma

wenyihu6 committed:
kvserver: make default multi-metric for testing

wenyihu6 commented:
Rebased again on master - some of the rebase were not very trivial (especially when resolving conflicts with 21b454c6a7cc6386d4bd643e2a6daaa3165e5422 and https://github.com/cockroachdb/cockroach/blob/e1648f8c8b0db7facd4f2f27a7bb6ce71c010f10/pkg/kv/kvserver/asim/state/impl.go#L1161-L1174).

wenyihu6 committed:
mma: trivial fix for ./dev build errors

wenyihu6 committed:
kvserver: add more logs

sumeerbhola committed:
mma: provide StoreLeaseholderMsg when calling ComputeChanges

wenyihu6 committed:
kvserver: log the post apply changes in lq.as.PostApply

wenyihu6 committed:
mma: panic if removePendingChangeTracking fails to find changeID in rangeState.pendingChanges

sumeerbhola committed:
mma: do not require pending change to exist in AdjustPendingChangesDisposition

wenyihu6 committed:
kvserver: do not update local store pool during PostApply if failed

sumeerbhola committed:
mma: document consequence of pendingReplicaChange being for a single store

sumeerbhola committed:
mma: improve variable name (no behavior change)

sumeerbhola committed:
mma: fix build for cluster_state_test.go

sumeerbhola committed:
kvserver: improve consistency of TryConstructMMARangeMsg

sumeerbhola committed:
kvserver: when MMA produces change for unknown replica, let it know about failure to enact

sumeerbhola committed:
kvserver,mma: improvements to StoreLeaseholderMsg

- MMA's knowledge about stores is passed to the Replica constructing the
  RangeMsg. If there is a replica on a store that is not known to MMA, the
  RangeMsg is not produced and such ignored ranges are counted and logged.
  This creates the illusion for MMA that a local store is not the
  leaseholder and will therefore not be involved in rebalancing it. This
  is a fix to the previous approach which dropped the RangeMsg after it
  had been produced by Replica, which is not permitted (since the changed
  logic in Replica relies on a message it produces to be fed to MMA).

- When a RangeMsg is produced with !Populated, the set of replicas is
  included anyway. This allows MMA to check if its state has diverged
  from the authoritative source, and correct itself. This is to work
  around a recent roachtest failure where MMA was told that a lease transfer
  was successful (by the mma store rebalancer) when it was not, which
  resulted in a permanent divergence of state.

wenyihu6 committed:
mma: print store leaseholder msg and store load msg better

sumeerbhola committed:
mma: document hazard with pendingReplicaChange

sumeerbhola committed:
mma: fix test breakage caused by a previous the semantic change to RangeMsg

The test broke due to commit
https://github.com/cockroachdb/cockroach/pull/142138/commits/fa2076ebbc8896fd267cc8aab773b6276ffcae7a

sumeerbhola committed:
mma: introduce rangeState.localRangeOwner for rangeState gc

sumeerbhola committed:
mma: introduce mma.StoreAttributesAndLocality and use that instead of StoreDescriptor

This is a cleanup since we only cared about certain fields in the StoreDescriptor.
MMA still accepts a StoreDescriptor in SetStore, which will be changed later.

sumeerbhola committed:
mma: remove remaining uses of StoreDescriptor

sumeerbhola committed:
mma: add a comment about improving locking

wenyihu6 committed:
mma: pre-check whether pending changes are still valid to apply/undo

sumeerbhola committed:
mma: fix some commentary related to ReplicaChange

wenyihu6 committed:
mma: fix preCheckOnUndoReplicaChanges to check a list of changes as well

Previously, preCheckOnUndoReplicaChanges validated each individual change separately. This was inadequate since the full set of changes is meant to be undone together - an intermediate states may be invalid even if the final state is valid. This commit updates the logic to apply and validate the full set of changes.

sumeerbhola committed:
mma: some cleanup for the preCheck* functions and callers

wenyihu6 committed:
mma: should panic if range is missing during pendingChangeEnacted

wenyihu6 committed:
mma: populate localRangeOwner correctly when initializing

sumeerbhola committed:
mma: fix some faulty logic around pending changes when processing a StoreLeaseholderMsg

wenyihu6 committed:
mma: add mma.external.success

sumeerbhola committed:
mma: log spam tweak

wenyihu6 committed:
mma: add more metrics

sumeerbhola committed:
mma,asim: improve context propagation and tags

Also commented out a log statement in clusterState.createPendingChanges
that was mostly redundant.

sumeerbhola committed:
mma: add rangeState.lastFailedChange

It is used by MMA to delay making another change to a range for 60s,
after a previous change needed to be undone.

wenyihu6 committed:
mma: fix KnownStores

wenyihu6 committed:
mma: add more logs (most are VInfo)

tbg committed:
kvserver: correctly tally WriteBytesPerSecond

We were actually summing up the write count per second instead. Found in
trivial skew workload experiment, where write values were way too low.

tbg committed:
mma: disable min{Replica,Lease}LoadFraction

See comment.

sumeerbhola committed:
mma: add comment for rangeState.lastFailedChange

sumeerbhola committed:
mma: comment about ignoring small ranges for rebalancing

sumeerbhola committed:
mma: add ReplicaCount to SecondaryLoadDimension

sumeerbhola committed:
mma: use the mean range load to decide the threshold of the top-k ranges

sumeerbhola committed:
asim: add mma_skewed_cpu_skewed_write_more_ranges

This does not rebalance sufficiently. Needs investigation.

wenyihu6 committed:
mma: skip maxFractionPendingIncrease if reported load is the same as adjusted load

Previously, MMA set ss.maxFractionPendingIncrease and
ss.maxFractionPendingDecrease to 1000 when the reported load was 0. This could
occur when some dimension (e.g. write bandwidth during read-only workloads).
This patch skips the adjustment when both reported and adjusted loads are
equal for a dimension. Note that if the reported load is
zero but the adjusted load is non-zero, we may still fall back to 1000.

wenyihu6 committed:
asim: add standard deviation calculation

This patch adds standard deviation to the store values output, measuring
variability along a single dimension. It reflects how evenly usage is
distributed across stores and serves as a simple metric to evaluate allocator
effectiveness over time and across versions.

tbg committed:
mma: add mma_one_voter_skewed_cpu_skewed_write

This converges, though at the end s2 still has `cpu=loadLow writes=overloadSlow`
while s1 has `cpu=loadNoChange writes=loadLow` but for some reason no more write load
is moving from s1 to s2.

I think it is because the writes also add CPU work, and s1 is
`cpu=loadNoChange`. So in this sense, this balanced entirely - nothing else can
be done.

tbg committed:
mma: add comment about node->store CPU capacity approach

wenyihu6 committed:
mma: clean up logs

wenyihu6 committed:
mma: clean up logs for ignore level and canShedAndAddLoad

tbg committed:
mma: add detailed explanation for topKReplicas

tbg committed:
mma: add logging counter

This is included as a tag and is available via an atomic load in the package,
making it convenient to look at "bunches" of log output that belong to the same
computation, and also makes it easy to set a breakpoint at a specific point in
time by adding a conditional for the desired mmaid.

tbg committed:
mma: comment on localStoreID

It's a little confusing that there is a distinction between the local store ID
on behalf of which rebalanceStores was invoked, and any other local store IDs.
This seems like the kind of thing that could lead to misunderstandings and/or
bugs.

tbg committed:
mma: assert on leaseholder status when considering lease shedding

A topKRange for a local store that has no pending changes must be the
leaseholder. Add an assertion for that and also a question - this doesn't seem
like it would hold if ranges could have replicas on multiple local stores.

tbg committed:
mma: explain why leaseholder is added to lease transfer target candidate list

tbg committed:
mma: more questions on leaseholder assertions

tbg committed:
mma: add question-TODO on re-adding leaseholder to lease transfer candidate set

tbg committed:
mma: don't consider rebalance if there are no candidates

We just re-added ourselves, so if there is one candidate only, we know we have nowhere for the lease to go.

This reduces annoying logging from computations that eventually refuse to self-transfer the lease.

tbg committed:
mma: log beginning of replica shedding section

tbg committed:
mma: log ignoreLevel

tbg committed:
mma: add yet another questino

tbg committed:
mma: add a TODO

tbg committed:
mma: crlfmt

wenyihu6 committed:
mma: remove duplicate boolean check overloadedDimPermitsChange

tbg committed:
mma: fix topK example

wenyihu6 committed:
mma: log when skipping lease shedding phase entirely

wenyihu6 committed:
mma: ./dev gen bazel files

wenyihu6 committed:
mma: fix ./dev lint

wenyihu6 committed:
mma: add ignoreLevel to redactcheck.go

wenyihu6 committed:
mma: update cluster state data driven test

Previously, we changed the log lines to be more explicit and this changed the output of the data-driven tests as well.

wenyihu6 committed:
mma: pass runtime load monitor to kvserver.NewStores

wenyihu6 committed:
kvserver: allow allocator sync to be nil when being passed in from tests

Tests can pass nil as the allocator sync when creating
newReplicateQueue. Replicate queue would simply do
nothing with allocator sync and not registering external
changes with mma. This is just a hack for the prototype as
we don't expect any unit tests to be using mma yet.

wenyihu6 committed:
kvserver: remove allocator sync field from StoreRebalancer

This patch removes the allocator sync field from StoreRebalancer.
It was left unused everywhere. I imagine we would swap
StoreRebalancer out with MMA entirely and not in a mixed state.
Maybe we populated it so that we can switch between the old
store rebalancer and mma? But mma should see the new cluster state
without any input from the old one?

wenyihu6 committed:
localtestcluster: ./dev gen

wenyihu6 committed:
localtestcluster: populate stopper in kvRuntimeLoadMonitor.Start correctly

wenyihu6 committed:
mma: fix TestLint/TestStaticCheck

wenyihu6 committed:
mma: log skipping lease shedding

wenyihu6 committed:
mma: add metric.Metadata properly for new metrics
